### DISCUSSION TITLE
Some modding questions

### DISCUSSION THREAD

**Discussion by Blackwolf7704** [SmartlyDressedGames/Unturned-3.x-Community#4319] [Q&A] [answered]
Continued from (https://github.com/SmartlyDressedGames/Unturned-3.x-Community/issues/4317)

The question addressed there is that changes in the master bundle cannot be reflected without restarting the game, and how to use AssetBundleCustomData.

However, when AssetBundleCustomData is added and uploaded, it does not work properly. (When workshopfileids and workshopfileid are specified at the same time, or when only fileids is specified, AssetBundleCustomData file is root directory)

Added inventory-related mod hooks to create unique items. I then created and applied a MonoBehavior script to process it, but the script cannot be found and does not work.

And scripts I found elsewhere don't seem to work either. (Example: https://github.com/olie304/UnturnedScripts)

How do I fix this?

**Comment by DanielWillett** [comment, marked as answer]
The way that script 'hack' works is by tricking unity into loading the script from the already in-game class instead of the one you defined. 

Most of the ones in that repo *should* work. If you want to double check you can either decompile the game or check this repository run by DiFFoZ: https://github.com/Unturned-Datamining/Unturned-Datamining (only the ones with `namespace SDG.Something` near the top of the file will show up in there though.)

You can define any field, method, or property in your version of the script that's defined in the original script and it should reference the in-game version when bundled.

Other than those, you just can't use custom scripts in mods, no way around it (other than using ModHooks and stuff).

As for asset bundle custom data, I was under the impression it worked but have never tested it so it's possible it's broken.

**Reply by Blackwolf7704** [reply]
Do I understand this correctly?

1. Except for ModHooks in the modpack (workshop item), creating other hooks or custom scripts will not work.

2. If you want it to work, make sure to load the script from a class in your game. (I am not sure about this even if I check the attached git repository.)

**Reply by Blackwolf7704** [reply]
So does loading a script in the game class mean I have to change the dll file in the game path?

**Reply by DrDallas1** [reply]
Absolutely not. Just use the script in Unity, and you have to do nothing else. A lot of olie's scripts work. You can only use scripts that exist within Unturned, not any custom ones.

**Reply by Blackwolf7704** [reply]
> Absolutely not. Just use the script in Unity, and you have to do nothing else. A lot of olie's scripts work. You can only use scripts that exist within Unturned, not any custom ones.

“You can only use scripts that exist within Unturned.” Can I understand this to mean redefining an existing function?

For example, is this the difference?

Part of the acid.cs code from the repository attached by DanielWillett
```
using System;
using UnityEngine;

namespace SDG.Unturned;

public class Acid: MonoBehaviour
{
     private bool isExploded;

     private Vector3 lastPos;

     publicGuid effectGuid;
.....
}
```

acid.cs code from olie's repository
```
using UnityEngine;

namespace SDG.Unturned
{
     // Plays the specified effect when something collides with the Collider component on the base Game Object
     // Destroys itself after one use and cannot be reset
     public class Acid: MonoBehaviour
     {
         public ushort effectID;
     }
}
```
Can I interpret this as extracting only part of the existing code?

**Reply by DanielWillett** [reply]
More like only exposing part of the code. You aren't rewriting anything, you're just telling Unity that those fields, properties, methods, etc will exist when they're loaded in-game.

Unity doesn't actually see whats in the methods or properties. it just sees:
```
Load new object
   Add SDG.Unturned.Acid component which has the following values:
     effectID = 15935;

or Load new object:
   Add SDG.Unturned.Flashbang component which has the following field:
     color = Color.green;
   Add SDG.Unturned.ActivationEventHook component which has the following event:
     OnEnabled, add event hook to this to Flashbang.Explode
```

**Reply by Blackwolf7704** [reply]
I'm not sure what it means to expose part of your code.

It feels like I'm overriding the value of a variable (field, property, method) within a function that exists in Unturned.

From the code comment you attached, I understand how the Load object works, but I'm not sure how to implement it.
